router.md

```diff
@@ -0,0 +1,38 @@
+
+
+## Pool Finder Design 
+A pool address can be calculated using `create2` logic by having 2 token addresses, factory address and `initCodeHash`, however the calculated pool address might not be deployed onchain, ie a direct pool for a token pair A/B is not available, so this brings in the usage of intermediate pools/tokens that will be used to be able to perform the swap between token A/B. 
+Intermediate tokens (called bases throughout this document) can be supplied at runtime or can have a default per chain as most univ2/v3 pools are deployed against well-known, highly adopted, commonly traded, etc tokens such as USDT, USDC, WETH, WBTC, DAI.
+
+So when trying to find pools for token pair A/B, the bases can be supplied as option (or use defaults), calculate the pool addresses for each combination and then check if the calculated addresses actually exist onchain or not, the ones that do exist will be stored into a struct field with their token details. the ones that do not exist, will be stored into the struct `blacklist` field, so next time there is new request for a new token pair, same process is repeated but before checking if the calculated addresses exist onchain or not, they are checked against the `blacklist` and those that already are known to exist.
+This will gurantee no same pool address is checked against onchain data twice during runtime as well as reducing the need for rpc calls as more pools are discovered.
+So they will be available throughout the `router` program execution once there is a request to find a route for a token pair.
+
+The `blacklist` needs to be purged once in while, the reason for this is that some pools might get deployed after they have gone into the blacklist, or there might have been some https/rpc error when a pool was initially checked that it exists onchain or not. 
+A simple logic can implemented to check the blacklist pools once in a while and if they happen to exist onchain, they will get removed from the blacklist, another suggestion can be to have a `pending blacklist`, ie not put pools directly into the blacklist for the first time they return error when read from onchain, but put them in a pending state, and if they happen to return error a few more times, then they will go into the blacklist.
+
+There should be methods to export/import the results, so the database of the found pools can be stored on disk and loaded from, to lower the rpc consumption as much as possible, after all once a pool is found, there is no need to find it again if the results are stored on a disk that can be loaded again at runtime.
+
+Since the database can grow overtime, it is logical to store them sorted, resulting in increasing performance on serach actions. the goal here is performance, so any logic or external library can be used to achieve this, not just limited to a sorted set/map.
```

Comment:
why does sorting improve the performance of a hash map?

Reply:
it isnt a hashmap, it is a btree map, so it is sorted by its own, I just meant to have a sorted record, so lets say there are thousands of pools, and a few need to be fetched from the records or checked if they exists or not, so having btree map vs a hashmap, I would go with btree map, does it make sense? @thedavidmeister